packages/block-library/src/gallery/test/index.native.js

```diff
@@ -27,102 +69,578 @@ afterAll( () => {
+	it( 'shows appender button when gallery has images', async () => {
+		const { galleryBlock, getByText } = await initializeWithGalleryBlock( {
+			numberOfItems: 1,
+			media,
+		} );
 
-	return screen;
-};
+		const appenderButton = within( galleryBlock ).getByA11yLabel(
+			/Gallery block\. Empty/
+		);
+		fireEvent.press( appenderButton );
 
-describe( 'Gallery block', () => {
-	it( 'inserts block', async () => {
-		const { getByA11yLabel } = await addGalleryBlock();
+		expect( getByText( 'Choose from device' ) ).toBeVisible();
+		expect( getByText( 'Take a Photo' ) ).toBeVisible();
+		expect( getByText( 'WordPress Media Library' ) ).toBeVisible();
+	} );
+
+	// This case is disabled until the issue (https://github.com/WordPress/gutenberg/issues/38444)
+	// is addressed.
+	it.skip( 'displays media options picker when selecting the block', async () => {
+		// Initialize with an empty gallery
+		const {
+			getByA11yLabel,
+			getByText,
+			getByTestId,
+		} = await initializeEditor( {
+			initialHtml: generateGalleryBlock( 0 ),
+		} );
+
+		// Tap on Gallery block
+		fireEvent.press( getByText( 'ADD MEDIA' ) );
+
+		// Observe that media options picker is displayed
+		expect( getByText( 'Choose images' ) ).toBeVisible();
+		expect( getByText( 'WordPress Media Library' ) ).toBeVisible();
+
+		// Dimiss the picker
+		if ( Platform.isIOS ) {
+			fireEvent.press( getByText( 'Cancel' ) );
+		} else {
+			fireEvent( getByTestId( 'media-options-picker' ), 'backdropPress' );
+		}
 
-		const galleryBlock = await waitFor( () =>
-			getByA11yLabel( /Gallery Block\. Row 1/ )
+		// Observe that the block is selected, this is done by checking if the block settings
+		// button is visible
+		const blockActionsButton = getByA11yLabel( /Open Block Actions Menu/ );
+		expect( blockActionsButton ).toBeVisible();
+	} );
+
+	// Test case related to TC001 - Close/Re-open post with an ongoing image upload
+	// Reference: https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/gallery.md#tc001
+	it( 'finishes pending uploads upon opening the editor', async () => {
+		const { notifyUploadingState, notifySucceedState } = setupMediaUpload();
+
+		// Initialize with a gallery that contains two items that are being uploaded
+		const { galleryBlock } = await initializeWithGalleryBlock( {
+			numberOfItems: 2,
+			media,
+			useLocalUrl: true,
+		} );
+
+		// Notify that the media items are uploading
+		await notifyUploadingState( media[ 0 ] );
+		await notifyUploadingState( media[ 1 ] );
+
+		// Check that images are showing a loading state
+		expect(
+			within( getGalleryItem( galleryBlock, 1 ) ).getByTestId( 'spinner' )
+		).toBeVisible();
+		expect(
+			within( getGalleryItem( galleryBlock, 2 ) ).getByTestId( 'spinner' )
+		).toBeVisible();
+
+		// Notify that the media items upload succeeded
+		await notifySucceedState( media[ 0 ] );
+		await notifySucceedState( media[ 1 ] );
+
+		expect( getEditorHtml() ).toMatchSnapshot();
```

Comment:
**Preface:** Thank you for adding all of these tests for the Gallery Block. ❤️ 

I'm concerned with the inclusion of snapshot testing as I've only ever seen them be more troublesome than the value the allegedly provide. When a snapshot test fails, it's generally fairly difficult to understand if the snapshot assertion failed for the right or wrong reasons, e.g.,
- the test assertion for the subject under test legitimately failed
- some other nested component implementation detail indirectly changed the snapshot

, especially when you weren't the "original code owner."

In my previous environment, many times this resulted in the snapshots being improperly updated to reflect invalid new assertions that weren't in line with the expectations of the original test assertion. I'm not sure that that would happen here, but it's just given me this "PTSD" of "if we're using `toMatchSnapshot` we're probably not asserting correctly."

What do you think?

Reply:
I totally agree and to be honest, I'm not a big fan of snapshot testing for the same arguments you mentioned. However, I'd like to note that in this case the snapshots are generated from the HTML output of the editor, not the classic React tree, so I don't initially foresee those problems.

In fact, I decided to go this way in order to simplify the tests and make them smaller, when we want to check if the HTML output of the editor is the expected value. To clarify this, let me share an example:

```js
it( 'disables crop images setting (using snapshot testing)', async () => {
	// Initialize with a gallery that contains one item
	const screen = await initializeWithGalleryBlock( {
		numberOfItems: 1,
		media,
	} );
	const { getByText } = screen;

	await openBlockSettings( screen );

	// Disable crop images setting
	fireEvent.press( getByText( 'Crop images' ) );
	expect( getEditorHtml() ).toMatchSnapshot();
} );

it( 'disables crop images setting (not using snapshot testing)', async () => {
	const expectedHTML = `<!-- wp:gallery {"imageCrop":false,"linkTo":"none"} -->
	<figure class="wp-block-gallery has-nested-images columns-default"><!-- wp:image {"id":2000} -->
	<figure class="wp-block-image"><img src="https://test-site.files.wordpress.com/local-image-1.jpeg" alt="" class="wp-image-2000"/></figure>
	<!-- /wp:image --></figure>
	<!-- /wp:gallery -->`;

	// Initialize with a gallery that contains one item
	const screen = await initializeWithGalleryBlock( {
		numberOfItems: 1,
		media,
	} );
	const { getByText } = screen;

	await openBlockSettings( screen );

	// Disable crop images setting
	fireEvent.press( getByText( 'Crop images' ) );
	expect( getEditorHtml() ).toBe( expectedHTML );
} );
```

Reply:
That does make me feel better about the odds that the snapshot serialization changes inadvertently when child component implementation details change. However, is there a reason that it wouldn't still be possible even if it was simply the serialization of the HTML output of the editor? I was trying to deep dive into the `subscribeParentGetHtml` implementation that interacts through the bridge with the `requestGetHtml` event to get a feel for the details. So far all files I've looked into have `.native` suffixes or are in the native layer so it doesn't appear that the core `gutenberg` code appears to be involved much, but I'm not confident.

That being said, I don't feel as strongly against using these types of snapshots. Even if it was possible for these to change inadvertently, I don't imagine the serialization of each block type changes fairly often. The most that would likely change would be default metadata in the HTML comments, and that would likely be a valid reason for an assertion to fail.

Thanks for elaborating further. I appreciate it. I am curious as to your thoughts about my question above so I won't resolve the conversation at the moment just so it doesn't get lost, but I intend to resolve the conversation thread.

Reply:
> However, is there a reason that it wouldn't still be possible even if it was simply the serialization of the HTML output of the editor

Of course, it would be still possible but I foresee similar odds as if we place the HTML expectation within the test code. Eventually, we might have a fixture file to include them, but I think it would be turn out to be quite similar to what we have currently in the snapshot file. So I don't have a strong opinion on whether to include it, to be honest, I'm open to both.